mlflow/tensorflow/__init__.py

```diff
@@ -769,6 +787,39 @@ def train(original, self, *args, **kwargs):
 
         result = original(self, *args, **kwargs)
 
+        def _extract_data_from_input_fn():
+            def _extract_dict_or_tensor(next_input):
+                input_ex_dict_tensor = None
+                if isinstance(next_input, dict):
+                    input_ex_dict_tensor = {
+                        k: _extract_tensor_as_numpy_input_example_slice(v)
+                        for k, v in next_input.items()
+                    }
+                elif isinstance(next_input, tensorflow.Tensor):
+                    input_ex_dict_tensor = _extract_tensor_as_numpy_input_example_slice(next_input)
+                return input_ex_dict_tensor
+
+            input_training_data = kwargs.get("input_fn")()
+            input_features = None
+            if isinstance(input_training_data, tuple):
+                input_features = _extract_dict_or_tensor(input_training_data[0])
+            elif isinstance(input_training_data, tensorflow.data.Dataset):
+                train_iter = iter(input_training_data)
+                next_input = next(train_iter)[0]
+                input_features = _extract_dict_or_tensor(next_input)
```

Comment:
Can we add some inline comments that explain what the data looks like for each type, including a small example?

This will help with maintainability; for example, it will make it easier to understand why `next_input = next(train_iter)[0]`. For my own understanding, why do we take the first element of the resulting output and discard the others?